subgraphs/osmosis/src/modules/Swap.ts

```diff
@@ -0,0 +1,347 @@
+import { BigDecimal, BigInt, cosmos, log } from "@graphprotocol/graph-ts";
+import {
+  Swap as SwapTransaction,
+  LiquidityPool as LiquidityPoolStore,
+  Token,
+} from "../../generated/schema";
+import * as constants from "../common/constants";
+import {
+  getOrCreateDexAmmProtocol,
+  getOrCreateToken,
+} from "../common/initializer";
+import * as utils from "../common/utils";
+import {
+  updateMetrics,
+  updateSnapshotsVolume,
+  updateSupplySideRevenue,
+  updateTokenVolumeAndBalance,
+} from "./Metrics";
+
+function createSwapTransaction(
+  from: string,
+  liquidityPool: LiquidityPoolStore,
+  transaction: cosmos.TxResult,
+  block: cosmos.HeaderOnlyBlock,
+  inputToken: string,
+  inputTokenAmount: BigInt,
+  amountInUSD: BigDecimal,
+  outputToken: string,
+  outputTokenAmount: BigInt,
+  amountOutUSD: BigDecimal
+): void {
+  if (!transaction) {
+    return;
+  }
+  const transactionId = "swap-" + transaction.hash.toHexString();
```

Comment:
The ID also needs the log index!

Reply:
Osmosis and cosmos logging mechanism is somewhat different with EVM chain. Every message (action) is mapped with a transaction which all includes the related logs. So with a transaction hash,  one can find the related message details and log. Here is the details for a swap transaction hash https://bigdipper.live/osmosis/transactions/EBE7B641DDFC13647F70C16E6B754D7E6BA868E3C8CFD17EE80B371B025BBCD9